File: clients/shared/VertexCore/Sources/VertexCore/Transport/MQTTPacketCodec.swift
import Foundation

/// MQTT 5.0 binary packet encoder/decoder for the minimal VPN subset.
///
/// Supports: CONNECT, CONNACK, PUBLISH (QoS 0), SUBSCRIBE (QoS 0),
/// SUBACK, PINGREQ, PINGRESP, DISCONNECT.
enum MQTTPacketCodec {

    // MARK: - Encode

    /// Encode a CONNECT packet (MQTT 5.0).
    static func encodeConnect(_ pkt: MQTTConnectPacket) -> Data {
        // Variable header
        var vh = Data()
        // Protocol Name: "MQTT"
        appendUTF8String("MQTT", to: &vh)
        // Protocol Version: 5
        vh.append(5)
        // Connect Flags
        var flags: UInt8 = 0
        if pkt.cleanStart { flags |= 0x02 }
        if pkt.username != nil { flags |= 0x80 }
        if pkt.password != nil { flags |= 0x40 }
        vh.append(flags)
        // Keep Alive
        appendUInt16(pkt.keepAlive, to: &vh)

        // Properties
        var props = Data()
        if let sei = pkt.sessionExpiryInterval {
            props.append(MQTTPropertyID.sessionExpiryInterval.rawValue)
            appendUInt32(sei, to: &props)
        }
        appendVariableInt(UInt32(props.count), to: &vh)
        vh.append(props)

        // Payload
        var payload = Data()
        appendUTF8String(pkt.clientID, to: &payload)
        if let u = pkt.username { appendUTF8String(u, to: &payload) }
        if let p = pkt.password { appendUTF8String(p, to: &payload) }

        return buildPacket(type: .connect, flags: 0, variableHeader: vh, payload: payload)
    }

    /// Encode a PUBLISH packet (QoS 0, no packet ID).
    static func encodePublish(_ pkt: MQTTPublishPacket) -> Data {
        var vh = Data()
        // Topic
        appendUTF8String(pkt.topic, to: &vh)
        // No packet ID for QoS 0

        // Properties
        var props = Data()
        if let expiry = pkt.messageExpiry {
            props.append(MQTTPropertyID.messageExpiryInterval.rawValue)
            appendUInt32(expiry, to: &props)
        }
        appendVariableInt(UInt32(props.count), to: &vh)
        vh.append(props)

        let flags: UInt8 = pkt.retain ? 0x01 : 0x00
        return buildPacket(type: .publish, flags: flags, variableHeader: vh, payload: pkt.payload)
    }

    /// Encode a SUBSCRIBE packet (QoS 0).
    static func encodeSubscribe(_ pkt: MQTTSubscribePacket) -> Data {
        var vh = Data()
        // Packet ID
        appendUInt16(pkt.packetID, to: &vh)
        // Properties (empty)
        vh.append(0) // property length = 0

        // Payload: topic filters with QoS 0
        var payload = Data()
        for topic in pkt.topics {
            appendUTF8String(topic, to: &payload)
            payload.append(0x00) // QoS 0, no options
        }

        return buildPacket(type: .subscribe, flags: 0x02, variableHeader: vh, payload: payload)
    }

    /// Encode a PINGREQ packet (2 bytes).
    static func encodePingReq() -> Data {
        Data([0xC0, 0x00]) // type=12, flags=0, remaining=0
    }

    /// Encode a DISCONNECT packet (reason=normal).
    static func encodeDisconnect() -> Data {
        // Type + flags, remaining length, reason code, property length
        Data([0xE0, 0x02, 0x00, 0x00])
    }

    // MARK: - Decode

    /// Try to decode one complete packet from the buffer.
    /// Returns (packet type, decoded data, bytes consumed) or nil if incomplete.
    static func tryDecode(_ buffer: Data) -> (MQTTPacketType, Data, Int)? {
        guard buffer.count >= 2 else { return nil }

        let firstByte = buffer[buffer.startIndex]
        let typeRaw = firstByte >> 4

        // Decode remaining length (variable-length integer)
        var multiplier: UInt32 = 1
        var remainingLength: UInt32 = 0
        var offset = buffer.startIndex + 1

        while offset < buffer.endIndex {
            let byte = buffer[offset]
            remainingLength += UInt32(byte & 0x7F) * multiplier
            offset += 1
            if byte & 0x80 == 0 { break }
            multiplier *= 128
            if multiplier > 128 * 128 * 128 { return nil } // max 4 bytes
        }

        let headerSize = offset - buffer.startIndex
        let totalSize = headerSize + Int(remainingLength)
        guard buffer.count >= totalSize else { return nil } // incomplete

        guard let packetType = MQTTPacketType(rawValue: typeRaw) else {
            return nil
        }

        let packetData = buffer[buffer.startIndex..<(buffer.startIndex + totalSize)]
        return (packetType, Data(packetData), totalSize)
    }

    /// Decode a CONNACK packet.
    static func decodeConnack(_ data: Data) throws -> MQTTConnackPacket {
        // Skip fixed header
        let (_, bodyOffset) = try parseFixedHeader(data)
        var offset = bodyOffset

        guard data.count >= offset + 2 else { throw MQTTCodecError.incompletePacket }

        // Acknowledge Flags
        let ackFlags = data[offset]
        let sessionPresent = (ackFlags & 0x01) != 0
        offset += 1

        // Reason Code
        let reasonCode = data[offset]
        offset += 1

        // Properties
        var serverKeepAlive: UInt16?
        if offset < data.count {
            let (propLen, propOffset) = try readVariableInt(data, from: offset)
            offset = propOffset
            let propEnd = offset + Int(propLen)

            while offset < propEnd && offset < data.count {
                let propID = data[offset]
                offset += 1
                switch propID {
                case MQTTPropertyID.serverKeepAlive.rawValue:
                    guard offset + 2 <= data.count else { break }
                    serverKeepAlive = readUInt16(data, at: offset)
                    offset += 2
                default:
                    // Skip unknown property — read its value based on type
                    offset = skipProperty(propID, in: data, at: offset)
                }
            }
        }

        return MQTTConnackPacket(
            sessionPresent: sessionPresent,
            reasonCode: reasonCode,
            serverKeepAlive: serverKeepAlive
        )
    }

    /// Decode a PUBLISH packet (incoming, QoS 0).
    static func decodePublish(_ data: Data) throws -> MQTTPublishReceived {
        let (_, bodyOffset) = try parseFixedHeader(data)
        var offset = bodyOffset

        // Topic
        let (topic, topicEnd) = try readUTF8String(data, from: offset)
        offset = topicEnd

        // No packet ID for QoS 0 (QoS is in fixed header flags bits 1-2)

        // Properties
        if offset < data.count {
            let (propLen, propOffset) = try readVariableInt(data, from: offset)
            offset = propOffset + Int(propLen) // skip all properties
        }

        // Payload = rest of packet
        let payload = data[offset...]
        return MQTTPublishReceived(topic: topic, payload: Data(payload))
    }

    /// Decode a SUBACK packet.
    static func decodeSuback(_ data: Data) throws -> MQTTSubackPacket {
        let (_, bodyOffset) = try parseFixedHeader(data)
        var offset = bodyOffset

        guard data.count >= offset + 2 else { throw MQTTCodecError.incompletePacket }

        // Packet ID
        let packetID = readUInt16(data, at: offset)
        offset += 2

        // Properties
        if offset < data.count {
            let (propLen, propOffset) = try readVariableInt(data, from: offset)
            offset = propOffset + Int(propLen)
        }

        // Reason codes (one per topic)
        let reasonCodes = Array(data[offset...])
        return MQTTSubackPacket(packetID: packetID, reasonCodes: reasonCodes)
    }

    // MARK: - Helpers (private)

    private static func buildPacket(
        type: MQTTPacketType,
        flags: UInt8,
        variableHeader: Data,
        payload: Data
    ) -> Data {
        let remaining = variableHeader.count + payload.count
        var packet = Data()
        packet.append((type.rawValue << 4) | (flags & 0x0F))
        appendVariableInt(UInt32(remaining), to: &packet)
        packet.append(variableHeader)
        packet.append(payload)
        return packet
    }

    private static func parseFixedHeader(_ data: Data) throws -> (remainingLength: UInt32, bodyOffset: Int) {
        guard data.count >= 2 else { throw MQTTCodecError.incompletePacket }
        let (remainingLength, bodyOffset) = try readVariableInt(data, from: data.startIndex + 1)
        return (remainingLength, bodyOffset)
    }

    // MARK: - Variable-length integer

    static func appendVariableInt(_ value: UInt32, to data: inout Data) {
        var v = value
        repeat {
            var byte = UInt8(v & 0x7F)
            v >>= 7
            if v > 0 { byte |= 0x80 }
            data.append(byte)
        } while v > 0
    }

    static func readVariableInt(_ data: Data, from start: Int) throws -> (UInt32, Int) {
        var multiplier: UInt32 = 1
        var value: UInt32 = 0
        var offset = start

        while offset < data.count {
            let byte = data[offset]
            value += UInt32(byte & 0x7F) * multiplier
            offset += 1
            if byte & 0x80 == 0 { return (value, offset) }
            multiplier *= 128
            if multiplier > 128 * 128 * 128 {
                throw MQTTCodecError.malformedRemainingLength
            }
        }
        throw MQTTCodecError.incompletePacket
    }

    // MARK: - UTF-8 strings

    private static func appendUTF8String(_ string: String, to data: inout Data) {
        let utf8 = Data(string.utf8)
        appendUInt16(UInt16(utf8.count), to: &data)
        data.append(utf8)
    }

    private static func readUTF8String(_ data: Data, from start: Int) throws -> (String, Int) {
        guard start + 2 <= data.count else { throw MQTTCodecError.incompletePacket }
        let length = Int(readUInt16(data, at: start))
        let strStart = start + 2
        guard strStart + length <= data.count else { throw MQTTCodecError.incompletePacket }
        guard let string = String(data: data[strStart..<(strStart + length)], encoding: .utf8) else {
            throw MQTTCodecError.malformedUTF8String
        }
        return (string, strStart + length)
    }

    // MARK: - Integer helpers

    private static func appendUInt16(_ value: UInt16, to data: inout Data) {
        data.append(UInt8(value >> 8))
        data.append(UInt8(value & 0xFF))
    }

    private static func appendUInt32(_ value: UInt32, to data: inout Data) {
        data.append(UInt8((value >> 24) & 0xFF))
        data.append(UInt8((value >> 16) & 0xFF))
        data.append(UInt8((value >> 8) & 0xFF))
        data.append(UInt8(value & 0xFF))
    }

    private static func readUInt16(_ data: Data, at offset: Int) -> UInt16 {
        UInt16(data[offset]) << 8 | UInt16(data[offset + 1])
    }

    /// Skip an unknown MQTT 5.0 property value based on property ID.
    private static func skipProperty(_ propID: UInt8, in data: Data, at offset: Int) -> Int {
        var off = offset
        switch propID {
        // 1-byte properties
        case 0x01, 0x17, 0x19, 0x24, 0x25, 0x28, 0x29, 0x2A:
            off += 1
        // 2-byte properties
        case 0x13, 0x21, 0x22, 0x23:
            off += 2
        // 4-byte properties
        case 0x02, 0x11, 0x18, 0x27:
            off += 4
        // UTF-8 string properties
        case 0x03, 0x08, 0x12, 0x15, 0x1A, 0x1C, 0x1F:
            if off + 2 <= data.count {
                let len = Int(readUInt16(data, at: off))
                off += 2 + len
            }
        // Binary data properties
        case 0x09, 0x16:
            if off + 2 <= data.count {
                let len = Int(readUInt16(data, at: off))
                off += 2 + len
            }
        // Variable byte integer properties
        case 0x0B:
            while off < data.count {
                let b = data[off]
                off += 1
                if b & 0x80 == 0 { break }
            }
        // User property (string pair)
        case 0x26:
            for _ in 0..<2 {
                if off + 2 <= data.count {
                    let len = Int(readUInt16(data, at: off))
                    off += 2 + len
                }
            }
        default:
            // Unknown property — can't determine size, bail
            off = data.count
        }
        return off
    }
}

File: clients/windows/src/Vertex.Core/Transport/MqttPacketCodec.cs
using System.Buffers.Binary;
using System.Text;

namespace Vertex.Core.Transport;

/// <summary>
/// MQTT 5.0 binary packet encoder/decoder for the minimal VPN subset.
/// Supported: CONNECT, CONNACK, PUBLISH (QoS 0), SUBSCRIBE (QoS 0),
/// SUBACK, PINGREQ, PINGRESP, DISCONNECT.
///
/// Wire format is byte-for-byte identical to Swift <c>MQTTPacketCodec</c>
/// (<c>clients/shared/VertexCore</c>) and Kotlin <c>MqttPacketCodec</c>
/// (<c>clients/android/.../core/mqtt</c>) — keep in sync if any of those
/// changes.
/// </summary>
public static class MqttPacketCodec
{
    // ---------- Encode ----------

    /// <summary>Encode a CONNECT packet (MQTT 5.0).</summary>
    public static byte[] EncodeConnect(ConnectPacket pkt)
    {
        var vh = new MemoryStream();

        // Protocol Name "MQTT", version 5.
        AppendUtf8String(vh, "MQTT");
        vh.WriteByte(5);

        // Connect flags.
        byte flags = 0;
        if (pkt.CleanStart)        flags |= 0x02;
        if (pkt.Username != null)  flags |= 0x80;
        if (pkt.Password != null)  flags |= 0x40;
        vh.WriteByte(flags);

        // Keep Alive (UInt16 BE).
        AppendUInt16(vh, pkt.KeepAlive);

        // Properties.
        var props = new MemoryStream();
        if (pkt.SessionExpiryInterval is uint sei)
        {
            props.WriteByte(MqttPropertyId.SessionExpiryInterval);
            AppendUInt32(props, sei);
        }
        AppendVariableInt(vh, (uint)props.Length);
        props.Position = 0;
        props.CopyTo(vh);

        // Payload.
        var payload = new MemoryStream();
        AppendUtf8String(payload, pkt.ClientId);
        if (pkt.Username is { } u) AppendUtf8String(payload, u);
        if (pkt.Password is { } p) AppendUtf8String(payload, p);

        return BuildPacket(MqttPacketType.Connect, flags: 0, vh.ToArray(), payload.ToArray());
    }

    /// <summary>Encode a PUBLISH packet (QoS 0, no packet ID).</summary>
    public static byte[] EncodePublish(PublishPacket pkt)
    {
        var vh = new MemoryStream();
        AppendUtf8String(vh, pkt.Topic);

        var props = new MemoryStream();
        if (pkt.MessageExpirySeconds is uint expiry)
        {
            props.WriteByte(MqttPropertyId.MessageExpiryInterval);
            AppendUInt32(props, expiry);
        }
        AppendVariableInt(vh, (uint)props.Length);
        props.Position = 0;
        props.CopyTo(vh);

        byte flags = pkt.Retain ? (byte)0x01 : (byte)0x00;
        return BuildPacket(MqttPacketType.Publish, flags, vh.ToArray(), pkt.Payload);
    }

    /// <summary>Encode a SUBSCRIBE packet (QoS 0). MQTT 5.0 mandates fixed-header flags=0x02 here.</summary>
    public static byte[] EncodeSubscribe(SubscribePacket pkt)
    {
        // MQTT 5.0 §3.8.3 — at least one Topic Filter is required; a
        // payload-less SUBSCRIBE is a protocol error and the broker will
        // disconnect us. Catch the misuse here rather than over the wire.
        if (pkt.Topics.Count == 0)
        {
            throw new ArgumentException("SUBSCRIBE requires at least one topic filter.", nameof(pkt));
        }

        var vh = new MemoryStream();
        AppendUInt16(vh, pkt.PacketId);
        vh.WriteByte(0x00); // properties length = 0

        var payload = new MemoryStream();
        foreach (var topic in pkt.Topics)
        {
            AppendUtf8String(payload, topic);
            payload.WriteByte(0x00); // QoS 0, no options
        }

        return BuildPacket(MqttPacketType.Subscribe, flags: 0x02, vh.ToArray(), payload.ToArray());
    }

    /// <summary>PINGREQ — 2 bytes total: <c>0xC0 0x00</c>.</summary>
    public static byte[] EncodePingReq() => new byte[] { 0xC0, 0x00 };

    /// <summary>DISCONNECT (reason=normal): <c>0xE0 0x02 0x00 0x00</c>.</summary>
    public static byte[] EncodeDisconnect() => new byte[] { 0xE0, 0x02, 0x00, 0x00 };

    // ---------- Decode ----------

    /// <summary>
    /// Try to decode one complete packet from the buffer. Returns
    /// <c>(type, packetSlice, bytesConsumed)</c> or <c>null</c> if the
    /// buffer is incomplete.
    /// </summary>
    public static (MqttPacketType Type, byte[] Packet, int Consumed)? TryDecode(ReadOnlySpan<byte> buffer)
    {
        if (buffer.Length < 2) return null;

        byte firstByte = buffer[0];
        byte typeRaw = (byte)(firstByte >> 4);

        // Variable-length remaining length.
        uint multiplier = 1;
        uint remainingLength = 0;
        int offset = 1;

        while (offset < buffer.Length)
        {
            byte b = buffer[offset];
            remainingLength += (uint)(b & 0x7F) * multiplier;
            offset++;
            if ((b & 0x80) == 0) break;
            multiplier *= 128;
            if (multiplier > 128 * 128 * 128) return null; // max 4 bytes
            if (offset >= buffer.Length) return null;       // need more bytes
        }

        // The loop guarantees `offset >= 2` here (we only enter once `buffer.Length >= 2`,
        // and the loop body always increments offset before either breaking or returning).
        int headerSize = offset;
        int totalSize = headerSize + (int)remainingLength;
        if (buffer.Length < totalSize) return null;

        if (!Enum.IsDefined(typeof(MqttPacketType), typeRaw)) return null;

        return ((MqttPacketType)typeRaw, buffer[..totalSize].ToArray(), totalSize);
    }

    /// <summary>Decode a CONNACK packet.</summary>
    public static ConnackPacket DecodeConnack(ReadOnlySpan<byte> data)
    {
        int offset = ParseFixedHeader(data);

        if (data.Length < offset + 2) throw new MqttCodecException("CONNACK truncated.");

        bool sessionPresent = (data[offset] & 0x01) != 0;
        offset++;

        byte reasonCode = data[offset];
        offset++;

        ushort? serverKeepAlive = null;
        if (offset < data.Length)
        {
            (uint propLen, int propOffset) = ReadVariableInt(data, offset);
            offset = propOffset;
            // Clamp propEnd against data.Length so a malformed huge propLen
            // can't poison the loop bound (ReadVariableInt already caps the
            // value at ≤268M which fits in int, but clamp for clarity).
            int propEnd = (int)Math.Min((long)offset + propLen, data.Length);

            while (offset < propEnd && offset < data.Length)
            {
                byte propId = data[offset];
                offset++;
                switch (propId)
                {
                    case MqttPropertyId.ServerKeepAlive:
                        if (offset + 2 > data.Length) break;
                        serverKeepAlive = BinaryPrimitives.ReadUInt16BigEndian(data[offset..]);
                        offset += 2;
                        break;
                    default:
                        offset = SkipProperty(propId, data, offset);
                        break;
                }
            }
        }

        return new ConnackPacket(sessionPresent, reasonCode, serverKeepAlive);
    }

    /// <summary>Decode a PUBLISH packet (incoming, QoS 0).</summary>
    public static PublishReceived DecodePublish(ReadOnlySpan<byte> data)
    {
        int offset = ParseFixedHeader(data);

        (string topic, int afterTopic) = ReadUtf8String(data, offset);
        offset = afterTopic;

        if (offset < data.Length)
        {
            (uint propLen, int afterPropLen) = ReadVariableInt(data, offset);
            offset = AdvancePastProperties(afterPropLen, propLen, data.Length, "PUBLISH");
        }

        byte[] payload = data[offset..].ToArray();
        return new PublishReceived(topic, payload);
    }

    /// <summary>Decode a SUBACK packet.</summary>
    public static SubackPacket DecodeSuback(ReadOnlySpan<byte> data)
    {
        int offset = ParseFixedHeader(data);

        if (data.Length < offset + 2) throw new MqttCodecException("SUBACK truncated.");

        ushort packetId = BinaryPrimitives.ReadUInt16BigEndian(data[offset..]);
        offset += 2;

        if (offset < data.Length)
        {
            (uint propLen, int afterPropLen) = ReadVariableInt(data, offset);
            offset = AdvancePastProperties(afterPropLen, propLen, data.Length, "SUBACK");
        }

        var reasonCodes = data[offset..].ToArray();
        return new SubackPacket(packetId, reasonCodes);
    }

    // ---------- Helpers ----------

    private static byte[] BuildPacket(MqttPacketType type, byte flags, byte[] vh, byte[] payload)
    {
        int remaining = vh.Length + payload.Length;
        var ms = new MemoryStream(2 + remaining);
        ms.WriteByte((byte)(((byte)type << 4) | (flags & 0x0F)));
        AppendVariableInt(ms, (uint)remaining);
        ms.Write(vh, 0, vh.Length);
        ms.Write(payload, 0, payload.Length);
        return ms.ToArray();
    }

    private static int ParseFixedHeader(ReadOnlySpan<byte> data)
    {
        if (data.Length < 2) throw new MqttCodecException("Packet too short.");
        (uint _, int bodyOffset) = ReadVariableInt(data, 1);
        return bodyOffset;
    }

    /// <summary>
    /// Skip a packet's variable-header property block, validating that the
    /// declared <paramref name="propLen"/> fits inside the remaining buffer.
    /// Returns the offset of the first byte after the properties.
    /// </summary>
    private static int AdvancePastProperties(int afterPropLen, uint propLen, int bufferLength, string packetName)
    {
        // ReadVariableInt caps the value at ≤ 0x0FFFFFFF (≤ 268M), so the
        // (long) widening below is just defence-in-depth against future codec
        // changes.
        long end = (long)afterPropLen + propLen;
        if (end > bufferLength)
        {
            throw new MqttCodecException(
                $"{packetName} properties length {propLen} overshoots remaining buffer ({bufferLength - afterPropLen} bytes).");
        }
        return (int)end;
    }

    /// <summary>MQTT variable-length integer encode (1-4 bytes).</summary>
    public static void AppendVariableInt(Stream sink, uint value)
    {
        do
        {
            byte b = (byte)(value & 0x7F);
            value >>= 7;
            if (value > 0) b |= 0x80;
            sink.WriteByte(b);
        }
        while (value > 0);
    }

    /// <summary>MQTT variable-length integer decode. Returns (value, offsetAfterField).</summary>
    public static (uint Value, int NextOffset) ReadVariableInt(ReadOnlySpan<byte> data, int start)
    {
        uint multiplier = 1;
        uint value = 0;
        int offset = start;

        while (offset < data.Length)
        {
            byte b = data[offset];
            value += (uint)(b & 0x7F) * multiplier;
            offset++;
            if ((b & 0x80) == 0) return (value, offset);
            multiplier *= 128;
            if (multiplier > 128 * 128 * 128)
            {
                throw new MqttCodecException("Malformed remaining length.");
            }
        }
        throw new MqttCodecException("Truncated variable-length integer.");
    }

    private static void AppendUtf8String(Stream sink, string s)
    {
        var bytes = Encoding.UTF8.GetBytes(s);
        AppendUInt16(sink, (ushort)bytes.Length);
        sink.Write(bytes, 0, bytes.Length);
    }

    private static (string Value, int NextOffset) ReadUtf8String(ReadOnlySpan<byte> data, int start)
    {
        if (start + 2 > data.Length) throw new MqttCodecException("UTF-8 string length missing.");
        int length = BinaryPrimitives.ReadUInt16BigEndian(data[start..]);
        int strStart = start + 2;
        if (strStart + length > data.Length) throw new MqttCodecException("UTF-8 string truncated.");
        string s = Encoding.UTF8.GetString(data.Slice(strStart, length));
        return (s, strStart + length);
    }

    private static void AppendUInt16(Stream sink, ushort value)
    {
        sink.WriteByte((byte)(value >> 8));
        sink.WriteByte((byte)(value & 0xFF));
    }

    private static void AppendUInt32(Stream sink, uint value)
    {
        sink.WriteByte((byte)((value >> 24) & 0xFF));
        sink.WriteByte((byte)((value >> 16) & 0xFF));
        sink.WriteByte((byte)((value >> 8)  & 0xFF));
        sink.WriteByte((byte)( value        & 0xFF));
    }

    /// <summary>Skip an unknown MQTT 5.0 property based on its identifier byte's known wire size.</summary>
    private static int SkipProperty(byte propId, ReadOnlySpan<byte> data, int offset)
    {
        int o = offset;
        switch (propId)
        {
            // 1-byte
            case 0x01: case 0x17: case 0x19: case 0x24: case 0x25:
            case 0x28: case 0x29: case 0x2A:
                o += 1; break;
            // 2-byte
            case 0x13: case 0x21: case 0x22: case 0x23:
                o += 2; break;
            // 4-byte
            case 0x02: case 0x11: case 0x18: case 0x27:
                o += 4; break;
            // UTF-8 string
            case 0x03: case 0x08: case 0x12: case 0x15: case 0x1A: case 0x1C: case 0x1F:
                if (o + 2 <= data.Length)
                {
                    int len = BinaryPrimitives.ReadUInt16BigEndian(data[o..]);
                    o += 2 + len;
                }
                break;
            // Binary data
            case 0x09: case 0x16:
                if (o + 2 <= data.Length)
                {
                    int len = BinaryPrimitives.ReadUInt16BigEndian(data[o..]);
                    o += 2 + len;
                }
                break;
            // Variable byte int
            case 0x0B:
                while (o < data.Length)
                {
                    byte b = data[o];
                    o++;
                    if ((b & 0x80) == 0) break;
                }
                break;
            // User property (string pair)
            case 0x26:
                for (int i = 0; i < 2; i++)
                {
                    if (o + 2 <= data.Length)
                    {
                        int len = BinaryPrimitives.ReadUInt16BigEndian(data[o..]);
                        o += 2 + len;
                    }
                }
                break;
            default:
                // Unknown property — can't determine size, bail.
                o = data.Length;
                break;
        }
        return o;
    }
}
